src/main/java/com/pusher/client/connection/websocket/WebSocketConnection.java

```diff
@@ -257,6 +264,39 @@ public void onClose(final int code, final String reason, final boolean remote) {
             return;
         }
 
+        //Reconnection logic
+        if(state == ConnectionState.CONNECTED || state == ConnectionState.CONNECTING){
+
+            if(reconnectAttempts < MAX_CONNECTION_ATTEMPTS){
+                tryReconnecting();
+            }
+            else{
+                updateState(ConnectionState.DISCONNECTING);
+                cancelTimeoutsAndTransitonToDisconnected();
+            }
+            return;
+        }
+
+        if (state == ConnectionState.DISCONNECTING){
+            cancelTimeoutsAndTransitonToDisconnected();
+        }
+    }
+
+    private void tryReconnecting() {
+        reconnectAttempts++;
+        updateState(ConnectionState.RECONNECTING);
```

Comment:
So, from what I understood, the state flow goes:
`Connected/Connecting` -> (`Reconnecting` -> `Connecting`)*6 -> `Disconnecting` -> `Disconnected`

Not sure if this would be clearer:
`Connected/Connecting` -> `Reconnecting`*6 -> `Disconnecting` -> `Disconnected`